File: connector-server/src/utils/prefixed-redis.ts
/**
 * Prefixed Redis client wrapper.
 *
 * 把 node-redis client (`createClient(...)`) 包一层 Proxy，让所有 key /
 * channel 接受/写出的位置透明地加 `<REDIS_KEY_PREFIX>:` 前缀。调用方
 * （route / service / worker）继续写裸 key：
 *
 *   await fastify.redis.get('roster:v2:list:123')    // 实际命中 sit:roster:v2:list:123
 *   await fastify.redis.set('lock:crew:42', token)   // 实际写入 sit:lock:crew:42
 *
 * 这把 v1 设计里"call site 必须自觉 withPrefix(...)"的契约变成不可绕过：
 * 不存在"忘了加 prefix"的漏网之鱼。
 *
 * BullMQ 的 ioredis 客户端**不**走这个 wrapper——它的 queue/worker 名字
 * 由 `withBullmqPrefix` 注入（`_` 分隔符），并且 BullMQ 自己的内部 key
 * 用 `bull:` 前缀，硬性不能被外面再加一层 `:`。BullMQ 的连接由
 * `plugins/bullmq.ts` 单独管。
 *
 * 设计取舍：每次调用都重新 `redisKeyPrefix()` 读 env（不缓存），与
 * `redis-key-prefix.ts` 一致——确保 vitest 在 `vi.resetModules()` 之间切
 * REDIS_KEY_PREFIX 仍然生效。
 */

import type { RedisClientType } from 'redis'
import { redisKeyPrefix } from './redis-key-prefix.js'

/** 当前 env 的 prefix 字符串（含末尾 `:`），空 prefix 退化到 `''`。 */
const pfx = (): string => {
  const p = redisKeyPrefix()
  return p ? `${p}:` : ''
}

/** 给一个 key 加 prefix。已带 prefix（首段 = prefix）的不重复加。 */
const prefixKey = (k: string): string => {
  const p = pfx()
  if (!p) return k
  if (k.startsWith(p)) return k
  return `${p}${k}`
}

/** 把多 key 数组逐个加 prefix。 */
const prefixKeys = (ks: readonly string[]): string[] => ks.map(prefixKey)

/** 列表形式的 eval keys（来自 options.keys）。 */
const prefixEvalKeys = (opts: unknown): unknown => {
  if (!opts || typeof opts !== 'object') return opts
  const o = opts as Record<string, unknown>
  if (Array.isArray(o.keys)) {
    return { ...o, keys: prefixKeys(o.keys as string[]) }
  }
  return opts
}

/** 转换 scan options 的 MATCH 模式。已带 prefix 的不重复加。 */
const prefixScanOpts = (opts: unknown): unknown => {
  if (!opts || typeof opts !== 'object') return opts
  const o = opts as Record<string, unknown>
  if (typeof o.MATCH === 'string') {
    return { ...o, MATCH: prefixKey(o.MATCH) }
  }
  return opts
}

/** 给一个 pubsub channel 加 prefix。 */
const prefixChannel = (ch: string): string => prefixKey(ch)

/**
 * node-redis v4 的 RedisClientType 有 ~100 个方法。这里只列出我们在
 * route / service / worker 里实际用到的 + 一些高危的近邻方法。其它方法
 * 通过 Proxy 透传（不动 key）。
 */
const HANDLED_METHODS = new Set([
  // 单 key
  'get', 'set', 'del', 'unlink', 'incr', 'decr', 'expire', 'pexpire',
  'ttl', 'pttl', 'persist', 'exists', 'type', 'rename', 'renameNX',
  'getDel', 'getSet', 'getEx', 'setEx', 'pSetEx', 'setNX', 'setXX',
  // 多 key
  'mGet', 'mSet', 'mSetNX',
  // hash
  'hGet', 'hSet', 'hDel', 'hGetAll', 'hIncrBy', 'hLen', 'hKeys', 'hVals',
  'hExists', 'hSetNX',
  // set
  'sAdd', 'sRem', 'sMembers', 'sIsMember', 'sCard', 'sPop', 'sInter',
  'sUnion', 'sDiff', 'sMove',
  // 模式
  'scan', 'scanStream',
  // 脚本
  'eval', 'evalSha',
  // pub-sub
  'publish', 'subscribe', 'pSubscribe', 'unsubscribe', 'pUnsubscribe',
])

/**
 * 创建一个透明加 prefix 的 Redis client proxy。
 *
 * 用法：
 *   const raw = createClient({ url: env.REDIS_URL })
 *   await raw.connect()
 *   const redis = createPrefixedRedis(raw)
 *   fastify.decorate('redis', redis)
 */
export const createPrefixedRedis = (raw: RedisClientType): RedisClientType => {
  return new Proxy(raw, {
    get(target, prop, receiver) {
      if (typeof prop !== 'string') {
        return Reflect.get(target, prop, receiver)
      }
      const value = Reflect.get(target, prop, receiver)
      if (typeof value !== 'function') {
        return value
      }
      // Non-handled methods (e.g. `duplicate`, `connect`, `on`, `quit`,
      // `sendCommand`) must run with the raw client as `this` so internal
      // private-field reads work. `duplicate()` in particular returns a fresh
      // client that internally reads from the original class — wrapping the
      // return value in our proxy would break that.
      if (!HANDLED_METHODS.has(prop)) {
        return value.bind(target)
      }
      // 关键方法：拦截，重写 key 位置
      return (...args: unknown[]) => {
        switch (prop) {
          // 单 key
          case 'get':
          case 'getDel':
          case 'getSet':
          case 'getEx':
          case 'incr':
          case 'decr':
          case 'expire':
          case 'pexpire':
          case 'ttl':
          case 'pttl':
          case 'persist':
          case 'exists':
          case 'type':
          case 'rename':
          case 'renameNX':
          case 'setEx':
          case 'pSetEx':
          case 'setNX':
          case 'setXX':
            return (value as (...a: unknown[]) => unknown).apply(target, [prefixKey(args[0] as string), ...args.slice(1)])

          // set: key + value + opts
          case 'set':
            return (value as (...a: unknown[]) => unknown).apply(target, [prefixKey(args[0] as string), args[1], ...args.slice(2)])

          // del / unlink: 多 key 形式
          case 'del':
          case 'unlink': {
            if (args.length === 1 && Array.isArray(args[0])) {
              return (value as (...a: unknown[]) => unknown).apply(target, [prefixKeys(args[0] as string[])])
            }
            return (value as (...a: unknown[]) => unknown).apply(target, [prefixKey(args[0] as string), ...args.slice(1)])
          }

          // mGet: 数组
          case 'mGet':
            return (value as (...a: unknown[]) => unknown).apply(target, [prefixKeys(args[0] as string[])])

          // mSet / mSetNX: [['k','v','k','v']] 或 [k1, v1, k2, v2]
          case 'mSet':
          case 'mSetNX': {
            if (args.length === 1 && Array.isArray(args[0])) {
              const flat = args[0] as string[]
              const out: string[] = []
              for (let i = 0; i < flat.length; i += 2) {
                out.push(prefixKey(flat[i] as string))
                out.push(flat[i + 1] as string)
              }
              return (value as (...a: unknown[]) => unknown).apply(target, [out])
            }
            const out: unknown[] = []
            for (let i = 0; i < args.length; i += 2) {
              out.push(prefixKey(args[i] as string))
              out.push(args[i + 1])
            }
            return (value as (...a: unknown[]) => unknown).apply(target, out)
          }

          // hash: [key, field, ...]
          case 'hGet':
          case 'hExists':
          case 'hDel':
          case 'hIncrBy':
          case 'hSet':
          case 'hSetNX':
          case 'hGetAll':
          case 'hKeys':
          case 'hVals':
          case 'hLen':
            return (value as (...a: unknown[]) => unknown).apply(target, [prefixKey(args[0] as string), ...args.slice(1)])

          // set 类型
          case 'sAdd':
          case 'sRem':
          case 'sMembers':
          case 'sIsMember':
          case 'sCard':
          case 'sPop':
          case 'sMove':
            return (value as (...a: unknown[]) => unknown).apply(target, [prefixKey(args[0] as string), ...args.slice(1)])
          case 'sInter':
          case 'sUnion':
          case 'sDiff':
            return (value as (...a: unknown[]) => unknown).apply(target, [prefixKeys(args[0] as string[])])

          // scan / scanStream
          case 'scan': {
            if (args.length === 1 && args[0] && typeof args[0] === 'object' && !Array.isArray(args[0])) {
              return (value as (...a: unknown[]) => unknown).apply(target, [prefixScanOpts(args[0])])
            }
            if (args.length >= 2) {
              return (value as (...a: unknown[]) => unknown).apply(target, [args[0], prefixScanOpts(args[1]), ...args.slice(2)])
            }
            return (value as (...a: unknown[]) => unknown).apply(target, args)
          }
          case 'scanStream': {
            if (args.length >= 1 && args[0] && typeof args[0] === 'object' && !Array.isArray(args[0])) {
              return (value as (...a: unknown[]) => unknown).apply(target, [prefixScanOpts(args[0]), ...args.slice(1)])
            }
            return (value as (...a: unknown[]) => unknown).apply(target, args)
          }

          // eval / evalSha: (script, options, ...args)
          case 'eval':
          case 'evalSha': {
            if (args.length >= 2 && args[1] && typeof args[1] === 'object' && !Array.isArray(args[1])) {
              return (value as (...a: unknown[]) => unknown).apply(target, [args[0], prefixEvalKeys(args[1]), ...args.slice(2)])
            }
            return (value as (...a: unknown[]) => unknown).apply(target, args)
          }

          // pub-sub
          case 'publish':
            return (value as (...a: unknown[]) => unknown).apply(target, [prefixChannel(args[0] as string), ...args.slice(1)])
          case 'subscribe':
          case 'pSubscribe':
          case 'unsubscribe':
          case 'pUnsubscribe': {
            if (args.length === 1 && Array.isArray(args[0])) {
              return (value as (...a: unknown[]) => unknown).apply(target, [(args[0] as string[]).map(prefixChannel)])
            }
            return (value as (...a: unknown[]) => unknown).apply(target, args.map((a) => typeof a === 'string' ? prefixChannel(a) : a))
          }

          default:
            return (value as (...a: unknown[]) => unknown).apply(target, args)
        }
      }
    },
  })
}

File: engine-server/src/utils/prefixed_redis.py
"""Prefixed Redis client wrapper for engine-server.

Mirrors the live-server / connector-server / pbs-server client-layer prefix
wrapper (``src/utils/prefixed-redis.ts``). Wraps a ``redis.Redis`` instance
so all key-accepting methods transparently receive ``<REDIS_KEY_PREFIX>:``
prepended keys. Call sites pass bare keys, exactly as they would to a raw
``redis.Redis``:

    await r.get("optimizer:task:abc")     # actually hits sit:optimizer:task:abc
    await r.sadd("optimizer:tasks:all", id)

The same wrapper also intercepts ``pipeline()`` to return a
``PrefixedPipeline`` so every command in a transaction gets the prefix too.

The engine-server doesn't use BullMQ, so no ``withBullmqPrefix`` analogue is
needed here. (BullMQ rejects ``:`` in queue names; that's a TS-only concern.)

Why a Python wrapper, not a redis-py monkey-patch? Because the existing
``redis-py`` API doesn't expose a connection-level key prefix hook, and
redis-py 5.x methods go through ``Redis.execute_command`` which we'd have to
patch on every instance. A thin object wrapper is the smallest change that
keeps the call sites clean and the test surface tight.

Design notes:

* ``redisKeyPrefix()`` is read on **every** call (no cache), so a
  ``monkeypatch.setenv("REDIS_KEY_PREFIX", ...)`` between tests takes
  effect without ``cache_clear``. Same rule as
  ``src/utils/redis_key_prefix.py``.
* Already-prefixed keys (first segment == current prefix) are passed
  through unchanged, so double-wrapping is a no-op.
* Empty prefix degrades to bare keys (no leading ``:``).
* Methods that don't take keys (``ping``, ``info``, ``acl_*``, ...) are
  passed through transparently via ``__getattr__`` falling back to the
  underlying client.
"""

from __future__ import annotations

from typing import Any, Iterable, List, Optional

import redis

from src.utils.redis_key_prefix import redis_key_prefix


def _pfx() -> str:
    """Current prefix with trailing ``:`` (or ``''`` if no prefix)."""
    p = redis_key_prefix()
    return f"{p}:" if p else ""


def _prefix_key(k: str) -> str:
    p = _pfx()
    if not p:
        return k
    if k.startswith(p):
        return k
    return f"{p}{k}"


def _prefix_keys(keys: Iterable[str]) -> List[str]:
    return [_prefix_key(k) for k in keys]


class PrefixedPipeline:
    """Wrap a ``redis.client.Pipeline`` so every queued command is prefixed.

    Mirrors the methods engine-server actually uses on a pipeline:
    ``setex / sadd / srem / delete / smembers / mget / publish / execute``.
    Other pipeline commands fall through to the underlying pipeline object.
    """

    def __init__(self, raw: "redis.client.Pipeline"):
        self._raw = raw

    # ── Single-key commands on the pipeline ─────────────────────────
    def setex(self, name: str, time: int, value: Any) -> "PrefixedPipeline":
        self._raw.setex(_prefix_key(name), time, value)
        return self

    def psetex(self, name: str, time_ms: int, value: Any) -> "PrefixedPipeline":
        self._raw.psetex(_prefix_key(name), time_ms, value)
        return self

    def set(self, name: str, value: Any, **kwargs: Any) -> "PrefixedPipeline":
        self._raw.set(_prefix_key(name), value, **kwargs)
        return self

    def get(self, name: str) -> "PrefixedPipeline":
        self._raw.get(_prefix_key(name))
        return self

    def delete(self, *names: Any) -> "PrefixedPipeline":
        if len(names) == 1 and isinstance(names[0], (list, tuple)):
            self._raw.delete(*_prefix_keys(names[0]))
        else:
            self._raw.delete(*_prefix_keys(names))
        return self

    def unlink(self, *names: Any) -> "PrefixedPipeline":
        if len(names) == 1 and isinstance(names[0], (list, tuple)):
            self._raw.unlink(*_prefix_keys(names[0]))
        else:
            self._raw.unlink(*_prefix_keys(names))
        return self

    def ttl(self, name: str) -> "PrefixedPipeline":
        self._raw.ttl(_prefix_key(name))
        return self

    def pttl(self, name: str) -> "PrefixedPipeline":
        self._raw.pttl(_prefix_key(name))
        return self

    # ── Set commands ────────────────────────────────────────────────
    def sadd(self, name: str, *values: Any) -> "PrefixedPipeline":
        self._raw.sadd(_prefix_key(name), *values)
        return self

    def srem(self, name: str, *values: Any) -> "PrefixedPipeline":
        self._raw.srem(_prefix_key(name), *values)
        return self

    def smembers(self, name: str) -> "PrefixedPipeline":
        self._raw.smembers(_prefix_key(name))
        return self

    # ── Hash commands ───────────────────────────────────────────────
    def hset(self, name: str, *args: Any, **kwargs: Any) -> "PrefixedPipeline":
        self._raw.hset(_prefix_key(name), *args, **kwargs)
        return self

    def hget(self, name: str, key: str) -> "PrefixedPipeline":
        self._raw.hget(_prefix_key(name), key)
        return self

    def hdel(self, name: str, *keys: str) -> "PrefixedPipeline":
        self._raw.hdel(_prefix_key(name), *keys)
        return self

    # ── Multi-key / list commands ───────────────────────────────────
    def mget(self, keys: Iterable[str], **kwargs: Any) -> "PrefixedPipeline":
        self._raw.mget(_prefix_keys(keys), **kwargs)
        return self

    # ── Pub/Sub ─────────────────────────────────────────────────────
    def publish(self, channel: str, message: Any, **kwargs: Any) -> "PrefixedPipeline":
        self._raw.publish(_prefix_key(channel), message, **kwargs)
        return self

    # ── Pipeline control ────────────────────────────────────────────
    def execute(self, raise_on_error: bool = True) -> List[Any]:
        return self._raw.execute(raise_on_error=raise_on_error)

    def __enter__(self) -> "PrefixedPipeline":
        self._raw.__enter__()
        return self

    def __exit__(self, *args: Any) -> Any:
        return self._raw.__exit__(*args)

    # Anything we don't explicitly intercept falls through to the raw
    # pipeline. This is important so callers (e.g. ``pipe.reset()``,
    # ``pipe.watch(...)``, ``pipe.multi()``) keep working.
    def __getattr__(self, item: str) -> Any:
        return getattr(self._raw, item)


class PrefixedRedis:
    """Wrap a ``redis.Redis`` instance so all key-accepting methods
    transparently prepend ``<REDIS_KEY_PREFIX>:``.

    Methods not in the intercepted set (``ping``, ``info``, ``acl_*``,
    ``scan_iter``, ...) fall through to the underlying client via
    ``__getattr__``.

    The wrapper is intentionally a thin object wrapper, not a subclass of
    ``redis.Redis`` — subclassing would force us to keep up with every
    redis-py release, while composition only needs to know about the
    methods we actually call.
    """

    def __init__(self, raw: "redis.Redis"):
        self._raw = raw

    # ── Single-key ──────────────────────────────────────────────────
    def get(self, name: str) -> Any:
        return self._raw.get(_prefix_key(name))

    def getdel(self, name: str) -> Any:
        return self._raw.getdel(_prefix_key(name))

    def set(self, name: str, value: Any, **kwargs: Any) -> Any:
        return self._raw.set(_prefix_key(name), value, **kwargs)

    def setex(self, name: str, time: int, value: Any) -> Any:
        return self._raw.setex(_prefix_key(name), time, value)

    def psetex(self, name: str, time_ms: int, value: Any) -> Any:
        return self._raw.psetex(_prefix_key(name), time_ms, value)

    def setnx(self, name: str, value: Any) -> Any:
        return self._raw.setnx(_prefix_key(name), value)

    def incr(self, name: str, amount: int = 1) -> Any:
        return self._raw.incr(_prefix_key(name), amount)

    def incrby(self, name: str, amount: int) -> Any:
        return self._raw.incrby(_prefix_key(name), amount)

    def incrbyfloat(self, name: str, amount: float) -> Any:
        return self._raw.incrbyfloat(_prefix_key(name), amount)

    def decr(self, name: str, amount: int = 1) -> Any:
        return self._raw.decr(_prefix_key(name), amount)

    def decrby(self, name: str, amount: int) -> Any:
        return self._raw.decrby(_prefix_key(name), amount)

    def expire(self, name: str, time: int, **kwargs: Any) -> Any:
        return self._raw.expire(_prefix_key(name), time, **kwargs)

    def pexpire(self, name: str, time_ms: int, **kwargs: Any) -> Any:
        return self._raw.pexpire(_prefix_key(name), time_ms, **kwargs)

    def expireat(self, name: str, when: int, **kwargs: Any) -> Any:
        return self._raw.expireat(_prefix_key(name), when, **kwargs)

    def pexpireat(self, name: str, when: int, **kwargs: Any) -> Any:
        return self._raw.pexpireat(_prefix_key(name), when, **kwargs)

    def ttl(self, name: str) -> int:
        return self._raw.ttl(_prefix_key(name))

    def pttl(self, name: str) -> int:
        return self._raw.pttl(_prefix_key(name))

    def persist(self, name: str) -> Any:
        return self._raw.persist(_prefix_key(name))

    def exists(self, name: str) -> Any:
        return self._raw.exists(_prefix_key(name))

    def type(self, name: str) -> Any:
        return self._raw.type(_prefix_key(name))

    def rename(self, src: str, dst: str) -> Any:
        return self._raw.rename(_prefix_key(src), _prefix_key(dst))

    def renamenx(self, src: str, dst: str) -> Any:
        return self._raw.renamenx(_prefix_key(src), _prefix_key(dst))

    # ── Multi-key ───────────────────────────────────────────────────
    def mget(self, keys: Iterable[str], **kwargs: Any) -> List[Any]:
        return self._raw.mget(_prefix_keys(keys), **kwargs)

    def mset(self, mapping: dict, **kwargs: Any) -> Any:
        # Every KEY side gets prefixed; values pass through unchanged.
        prefixed = {_prefix_key(k): v for k, v in mapping.items()}
        return self._raw.mset(prefixed, **kwargs)

    def delete(self, *names: Any) -> Any:
        if len(names) == 1 and isinstance(names[0], (list, tuple)):
            return self._raw.delete(*_prefix_keys(names[0]))
        return self._raw.delete(*_prefix_keys(names))

    def unlink(self, *names: Any) -> Any:
        if len(names) == 1 and isinstance(names[0], (list, tuple)):
            return self._raw.unlink(*_prefix_keys(names[0]))
        return self._raw.unlink(*_prefix_keys(names))

    # ── Hash ────────────────────────────────────────────────────────
    def hget(self, name: str, key: str) -> Any:
        return self._raw.hget(_prefix_key(name), key)

    def hset(self, name: str, *args: Any, **kwargs: Any) -> Any:
        return self._raw.hset(_prefix_key(name), *args, **kwargs)

    def hsetnx(self, name: str, key: str, value: Any) -> Any:
        return self._raw.hsetnx(_prefix_key(name), key, value)

    def hdel(self, name: str, *keys: str) -> Any:
        return self._raw.hdel(_prefix_key(name), *keys)

    def hgetall(self, name: str) -> Any:
        return self._raw.hgetall(_prefix_key(name))

    def hincrby(self, name: str, key: str, amount: int = 1) -> Any:
        return self._raw.hincrby(_prefix_key(name), key, amount)

    def hincrbyfloat(self, name: str, key: str, amount: float = 0.0) -> Any:
        return self._raw.hincrbyfloat(_prefix_key(name), key, amount)

    def hlen(self, name: str) -> int:
        return self._raw.hlen(_prefix_key(name))

    def hkeys(self, name: str) -> List[Any]:
        return self._raw.hkeys(_prefix_key(name))

    def hvals(self, name: str) -> List[Any]:
        return self._raw.hvals(_prefix_key(name))

    def hexists(self, name: str, key: str) -> bool:
        return self._raw.hexists(_prefix_key(name), key)

    def hmget(self, name: str, keys: Iterable[str]) -> List[Any]:
        return self._raw.hmget(_prefix_key(name), list(keys))

    def hmset(self, name: str, mapping: dict) -> Any:
        return self._raw.hmset(_prefix_key(name), mapping)

    # ── Set ─────────────────────────────────────────────────────────
    def sadd(self, name: str, *values: Any) -> Any:
        return self._raw.sadd(_prefix_key(name), *values)

    def srem(self, name: str, *values: Any) -> Any:
        return self._raw.srem(_prefix_key(name), *values)

    def smembers(self, name: str) -> set:
        return self._raw.smembers(_prefix_key(name))

    def sismember(self, name: str, value: Any) -> bool:
        return self._raw.sismember(_prefix_key(name), value)

    def scard(self, name: str) -> int:
        return self._raw.scard(_prefix_key(name))

    def spop(self, name: str, count: Optional[int] = None) -> Any:
        if count is None:
            return self._raw.spop(_prefix_key(name))
        return self._raw.spop(_prefix_key(name), count)

    def smove(self, src: str, dst: str, value: Any) -> bool:
        return self._raw.smove(_prefix_key(src), _prefix_key(dst), value)

    def sinter(self, keys: Iterable[str], **kwargs: Any) -> set:
        return self._raw.sinter(_prefix_keys(keys), **kwargs)

    def sunion(self, keys: Iterable[str], **kwargs: Any) -> set:
        return self._raw.sunion(_prefix_keys(keys), **kwargs)

    def sdiff(self, keys: Iterable[str], **kwargs: Any) -> set:
        return self._raw.sdiff(_prefix_keys(keys), **kwargs)

    # ── Pub/Sub ─────────────────────────────────────────────────────
    def publish(self, channel: str, message: Any, **kwargs: Any) -> int:
        return self._raw.publish(_prefix_key(channel), message, **kwargs)

    # ── Pipeline ────────────────────────────────────────────────────
    def pipeline(self, *args: Any, **kwargs: Any) -> PrefixedPipeline:
        raw_pipe = self._raw.pipeline(*args, **kwargs)
        return PrefixedPipeline(raw_pipe)

    # ── Passthrough for everything else ─────────────────────────────
    # ``ping``, ``info``, ``acl_*``, ``scan_iter``, ``config_*``, etc.
    # are exposed unchanged so callers can use the wrapper as a drop-in
    # for ``redis.Redis`` without losing any non-key methods.
    def __getattr__(self, item: str) -> Any:
        return getattr(self._raw, item)
